src/HttpLayer.hs

```diff
@@ -40,41 +42,62 @@ emptyHTTPNetworkLayer = HTTPNetworkLayer
 ------------------------------------------------------------
 
 -- | General api request function
-apiRequest :: Config -> Text -> Request
-apiRequest Config{..} u =
-    setRequestPath (encodeUtf8 path) $
-    addRequestHeader "Content-Type" "application/json" $
-    setRequestBasicAuth
-        (encodeUtf8 cfgEmail <> "/token")
-        (encodeUtf8 cfgToken) $
-    parseRequest_ (toString (cfgZendesk <> path))
+apiRequest
+    :: Config
+    -> Text
+    -> Either String Request
+apiRequest Config{..} u = mapLeft show $ catches buildRequest handlerList
   where
+    buildRequest :: forall m. MonadCatch m => m Request
+    buildRequest = do
+        req <- parseUrlThrow (toString (cfgZendesk <> path)) -- TODO(md): Get a list of exceptions that parseUrlThrow can throw
+        return $ setRequestPath (encodeUtf8 path) $
+                 addRequestHeader "Content-Type" "application/json" $
+                 setRequestBasicAuth
+                     (encodeUtf8 cfgEmail <> "/token")
+                     (encodeUtf8 cfgToken) req
+    handlerList :: MonadCatch m => [Handler m Request]
+    handlerList = [
+                    handlerJSON
+                  , handlerHTTP
+                  ]
+    handlerJSON :: MonadCatch m => Handler m Request
+    handlerJSON = Handler $ \(ex :: JSONException) -> error $ show ex
+    handlerHTTP :: MonadCatch m => Handler m Request
+    handlerHTTP = Handler $ \(ex :: HttpException) -> error $ show ex
     path :: Text
     path = "/api/v2" <> u
 
 -- | Api request but use absolute path
-apiRequestAbsolute :: Config -> Text -> Request
+apiRequestAbsolute
+    :: Config
+    -> Text
+    -> Either String Request
 apiRequestAbsolute Config{..} u =
-    addRequestHeader "Content-Type" "application/json" $
+    Right <$> addRequestHeader "Content-Type" "application/json" $
     setRequestBasicAuth (encodeUtf8 cfgEmail <> "/token") (encodeUtf8 cfgToken) $
     parseRequest_(toString u)
 
 -- | Request PUT
-addJsonBody :: forall a. ToJSON a => a -> Request -> Request
-addJsonBody body req = setRequestBodyJSON body $ setRequestMethod "PUT" req
+addJsonBody :: forall a. ToJSON a => a -> Request -> Either String Request
```

Comment:
As a general rule, if you just wrap `Either` with a single constructor (`Right`), you probably don't need it.
Can you rewrite this to catch any exceptions and in that case to return it?

Reply:
I agree with your observation. However, an exception can occur and that's why there is a call to the `catch` function.

Reply:
And where do you catch it?

Reply:
It's caught right there, on the first line of `addJsonBody`.

Reply:
And where in the type signature does it say that this can fail?

Reply:
@mdimjasevic Think about what I'm asking here, don't just answer back - what place is that exception coming from? How do you know that this function needs to address this exception? What kind of signature is on this function, and does that tell you something about its behavior?
Think about exceptions - do they occur on pure functions? In which cases? A lot of things to think about, no?

Reply:
All of these questions are easy to answer by peeking at the code. Functions from the HTTP library that are called in `addJsonBody` can throw exceptions, hence `addJsonBody` has a corresponding signature. But you must be right, these are great questions.